### ISSUE TITLE
[Question] Packet-level simulation using SMPI?

### ISSUE THREAD

**Issue by miharulidze** [simgrid/simgrid#336]
Dear SimGrid developers,

Recently, I read an article "Simulating MPI applications: the SMPI approach" which describes SimGrid framework.

Here is some points from my current understanding how SimGrid works:
1) SimGrid includes a high-level layer to intercept MPI calls (aka SMPI) from running application or app trace.
2) Then, SMPI calls SURF routines through SIMIX middle-ware. The SURF layer actually computes a time of communication routines using available flow-level/packet-level communication model.

In src/surf directory I found a lot of different modules, including smpi (actually described in article), ib, storage, ns-3 and etc.
Since I'm interested in hardware packet-level simulation of MPI workloads, here is my questions:
- Is it possible to use SMPI with ns-3 communication module?
- Are there easy way to verify that it works?

Thanks in advance!

**Comment by miharulidze** [comment]
I've done some tests and here are my current observations.
> Is it possible to use SMPI with ns-3 communication module?

It wasn't hard to build SimGrid with both SMPI and ns-3 modules. I just set up a corresponding Cmake options from documentation.
Also, the man-pages point out that we can pass a lot of options to `smpirun`, including `--cfg=network/model:ns-3`. I ran precompiled allreduce test from SMPI cmake tests with ns-3 dogbone platform file: `-platform examples/platforms/dogbone.xml` and `--cfg=network/model:ns-3` flag. 
It seemed to work OK. Allreduce results are the same to running without ns-3 option.
> Are there easy way to verify that it works?

I tried to set NS_LOG variable to `NS_LOG="*"`. Grep through the output log shows that there are some traffic flows around the nodes in ns-3. 

So, it really looks like smpi+ns-3 simulation actually works.
Correct me if i wrong, please.

**Comment by mquinson** [comment]
Hello, 

the best would be to look at the first lines of your simulation output. If the network model is switched to ns3, that should be written within the 5 to 10 first lines. I'm really short on time right now, but if you paste the head of your log, I'll tell you. 

Bye, Mt.

**Comment by alegrand** [comment]
Hello @miharulidze,

  and welcome to the SimGrid users community. Sorry for not answering earlier to your initial email on the simgrid-user mailing list. My answer would have been "In theory, it should work although I don't know whether anyone ever tried"and I'm happy to see that now I'll be able to answer "In practice, it seems to work as someone already did it!". ;-) Seriously, I'm very interest by any feedback you'll have on this path. When we implemented the simgrid/ns3 (and previously simgrid/gtnets) connection, our main goal was to study the validity of the SimGrid flow-level models. You may be interested in the [TOMACS 2013 article](https://hal.inria.fr/hal-00872476) we wrote on this topic. 

Although we have put a lot of efforts in validating SMPI/flow-models against real experiments, we never tried comparing against ns3 simulation. The TOMACS workload was a bunch of random flows as our goal was to cover many cases and detect weird corner cases. I can see two technical aspects which are likely to have an important impact on the realism of the simulations you will run with SMPI/ns3: 1) we rely on ns3 TCP connections so although it should be possible to hack SMPI/ns3 to use a different protocol, be careful with this as it may not be what you want 2) TCP is connection oriented and I believe that in our implementation of SimGrid/ns3, a TCP connection is opened every time two process engage into a communication. This is quite unrealistic as in real MPI implementations, connections are maintained throughout the whole lifespan of the program, which dramatically improves connection (no more slow-start).

In short, I'm very happy you're trying to go along this path and I'll be happy to discuss about this (although I'm very short in time like @mquinson and won't be able to help much on the ground). Feel free to ask questions and report us your investigations.

Best,

  Arnaud

**Comment by mquinson** [comment]
I confirm that we create one socket per communication. The entry point to start a communication (with this model) is [`NetworkNS3Model::communicate()`](https://framagit.org/simgrid/simgrid/blob/master/src/surf/network_ns3.cpp#L185). This jumps to [`new NetworkNS3Action()`](https://framagit.org/simgrid/simgrid/blob/master/src/surf/network_ns3.cpp#L310) that is where the ns-3 code is.

As you can see, we create a ns-3 source and sink + a socket for each communication. 

This could be improved, but we have to find a good match between the concepts of simgrid and the ones of ns-3. There is no socket in SimGrid that we could keep open, but we may do something around [receiving actors](https://simgrid.frama.io/simgrid/app_s4u.html#declaring-a-receiving-actor) of simgrid. 

In case you need these changes or something else, I fear we won't have the manpower to implement it for you, but  we'd be glad to guide you so that you that simgrid fits your needs.

HTH, Mt

**Comment by miharulidze** [comment]
Dear @mquinson @alegrand, 
 
Thank you for reply and advice!
I ran ping-pong example from SMPI tests and logs are at the end of this message (for some reason I cannot attach it as file on github).

It looks like during this tests SimGrid posts several 4 byte sends and ns-3 handles this requests.

> TCP is connection oriented and I believe that in our implementation of SimGrid/ns3, a TCP connection is opened every time two process engage into a communication. This is quite unrealistic as in real MPI implementations, connections are maintained throughout the whole lifespan of the program, which dramatically improves connection (no more slow-start).

I think, this is the question for future verification. MPICH and OpenMPI communication back-ends may use dynamic connection policy for QPs/sockets especially on a large scale in order to save node memory.

**Comment by miharulidze** [comment]
I ran p2p simulation with the following string: 
`smpirun -map -hostfile ./hostfile -platform ../../../examples/platforms/dogbone.xml -np 2 pt2pt-pingpong --log=root.:debug --cfg=network/model:ns-3`

In log I found several lines like that:

```
[C1:0:(1) 0.000000] /home/mikhail/Development/SimGrid-3.24/src/smpi/mpi/smpi_request.cpp:168: [smpi_request/VERBOSE] New send  request 0x5584b1da3440  [buf = 0x7f6e793c6b54, size = 4, src = 1, dst = 2, tag = 42, flags = 6]

[C1:0:(1) 0.000000] /home/mikhail/Development/SimGrid-3.24/src/smpi/mpi/smpi_request.cpp:443: [smpi_request/DEBUG] Send request 0x5584b1da3440 is detached

[C1:0:(1) 0.000000] /home/mikhail/Development/SimGrid-3.24/src/smpi/mpi/smpi_request.cpp:458: [smpi_request/DEBUG] buf 0x7f6e793c6b54 copied into 0x5584b1d05440

[C1:0:(1) 0.000000] /home/mikhail/Development/SimGrid-3.24/src/smpi/internals/smpi_host.cpp:62: [smpi_host/DEBUG] Searching for smpi/os: 4 is larger than the largest boundary, return 0.0000000000

[C1:0:(1) 0.000000] /home/mikhail/Development/SimGrid-3.24/src/simix/popping_bodies.cpp:34: [simix/DEBUG] Yield process '0' on simcall SIMCALL_COMM_ISEND (5)

[C1:0:(1) 0.000000] /home/mikhail/Development/SimGrid-3.24/src/kernel/actor/ActorImpl.cpp:280: [simix_process/DEBUG] Yield actor '0'

[C1:0:(1) 0.000000] /home/mikhail/Development/SimGrid-3.24/src/kernel/context/ContextSwapped.cpp:253: [simix_context/DEBUG] No more actors to run

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/simix/popping_generated.cpp:59: [simix_popping/DEBUG] Handling simcall 0x5584b1ce8858: SIMCALL_COMM_ISEND

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/kernel/activity/CommImpl.cpp:40: [simix_network/DEBUG] send from mailbox 0x5584b1dfb3b0

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/smpi/mpi/smpi_request.cpp:142: [smpi_request/DEBUG] Trying to match a send of src 1 against 1, tag 42 against 42, id -1 against -1

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/smpi/mpi/smpi_request.cpp:159: [smpi_request/DEBUG] match succeeded

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/smpi/mpi/smpi_request.cpp:115: [smpi_request/DEBUG] Trying to match a recv of src 1 against 1, tag 42 against 42, id -1 against -1

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/smpi/mpi/smpi_request.cpp:133: [smpi_request/DEBUG] match succeeded

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/kernel/activity/MailboxImpl.cpp:142: [simix_mailbox/DEBUG] Found a matching communication synchro 0x5584b1da3330

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/kernel/activity/CommImpl.cpp:69: [simix_network/DEBUG] Receive already pushed

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/surf/network_ns3.cpp:318: [ns3/DEBUG] Synchronizing NS-3 (time 0.000000) with SimGrid (time 0.000000)

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/surf/network_ns3.cpp:389: [ns3/DEBUG] Start simulator for at most 0.000000s (current time: 0.000000)

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/surf/network_ns3.cpp:391: [ns3/DEBUG] Simulator stopped at 0.000000s

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/surf/network_ns3.cpp:323: [ns3/DEBUG] Communicate from C1 to C2

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/surf/network_ns3.cpp:339: [ns3/DEBUG] ns3: Create flow of 4 Bytes from 2 to 4 with Interface 1.5.0.2

[0.000000] /home/mikhail/Development/SimGrid-3.24/src/kernel/activity/CommImpl.cpp:439: [simix_network/DEBUG] Starting communication 0x5584b1da3330 from 'C1' to 'C2' (surf_action: 0x5584b1da3600)
```

**Comment by mquinson** [comment]
Hello there, any news on this issue? @miharulidze, does SimGrid fit your needs after all? If so, can we close the bug, or is there maybe some part of this discussion that should be integrated to the documentation?
Thanks,

**Comment by miharulidze** [comment]
@mquinson Hello Martin and thanks for a feedback.

> does SimGrid fit your needs after all

At the moment, I have temporarily suspended testing/validation of SMPI+ns-3 functionality. At the same time, the idea to simulate the behaviour of hardware on packet-level on different MPI/HPC workloads still sounds very interesting to me, especially with SMPI, so I'll return to it in near future.
I mostly worried about the restriction on topology of system for ns-3 layer (routes cannot be longer than 1), because I need to simulate CLOS topology.

> maybe some part of this discussion that should be integrated to the documentation?

I think it's a good idea to clearly point out in documentation that it's _possible_ to use any low-level model like ns-3, flow-level or any other model in SURF with SMPI or any other layer.
For example, the publications as well as documentation are focused on flow-level model in context of SMPI and there is nothing about availability of other models in SURF. 
Sure, some combinations not very carefully validated and may give strange results, but the fact that we could use any combination of models is **very** cool! :-)

**Comment by mquinson** [comment]
I added a paragraph at the end of https://simgrid.frama.io/simgrid/app_smpi.html#simulating-your-code so I guess it is enough to close this bug. If not, feel free to reopen.
